File: Day20-25/25.c
#include <stdio.h>
#include <string.h>

#define TAPE_SIZE 15000
#define NUM_STEPS 12994925

typedef struct {
    int new_val;
    int movement;
    char next_state;
} transition;


transition turing_machine[6][2] = {
    {{1, +1, 'b'}, {0, -1, 'f'}}, // a
    {{0, +1, 'c'}, {0, +1, 'd'}}, // b
    {{1, -1, 'd'}, {1, +1, 'e'}}, // c
    {{0, -1, 'e'}, {0, -1, 'd'}}, // d
    {{0, +1, 'a'}, {1, +1, 'c'}}, // e
    {{1, -1, 'a'}, {1, +1, 'a'}}, // f
};

int runTM(int steps) {
    int tape[TAPE_SIZE] = {0};
    int loc = TAPE_SIZE/2;
    char state = 'a';
    int num_ones = 0;

    for(int i = 0; i < steps; i++) {
        int val = tape[loc];
        transition t = turing_machine[state - 'a'][val];

        if(t.new_val == 1 && val == 0) {
            num_ones++;
        } else if(t.new_val == 0 && val == 1) {
            num_ones--;
        }

        tape[loc] = t.new_val;
        loc += t.movement;
        state = t.next_state;
    }
    return num_ones;
} 

int main(){
    printf("Part 1: %d\n", runTM(12994925));
    printf("Part 2: ⭐\n");
}

File: Day20-25/25.go
package main

import "fmt"

type Transition struct {
	newVal    int
	movement  int
	nextState byte
}

var turingMachine = [][]Transition{
	// a
	{
		Transition{1, +1, 'b'},
		Transition{0, -1, 'f'}},
	// b
	{
		Transition{0, +1, 'c'},
		Transition{0, +1, 'd'}},
	// c
	{
		Transition{1, -1, 'd'},
		Transition{1, +1, 'e'}},
	// d
	{
		Transition{0, -1, 'e'},
		Transition{0, -1, 'd'}},
	// e
	{
		Transition{0, +1, 'a'},
		Transition{1, +1, 'c'}},
	// f
	{
		Transition{1, -1, 'a'},
		Transition{1, +1, 'a'}},
}

func runTM(steps int) int {
	tape := make([]int, 15000)
	loc := len(tape) / 2
	state := byte('a')
	numOnes := 0

	for i := 0; i < steps; i++ {
		val := tape[loc]
		t := turingMachine[state-'a'][val]

		if t.newVal == 1 && val == 0 {
			numOnes++
		} else if t.newVal == 0 && val == 1 {
			numOnes--
		}

		tape[loc] = t.newVal
		loc += t.movement
		state = t.nextState
	}
	return numOnes
}

func main() {
	steps := 12994925
	fmt.Printf("Part 1: %d\n", runTM(steps))
	fmt.Printf("Part 2: ⭐\n")
}
